sumdbaudit/witness/cmd/feeder/main.go

```diff
@@ -0,0 +1,295 @@
+	if wcpRaw, err := w.GetLatestCheckpoint(ctx, *logID); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			glog.Exitf("Failed to get witness checkpoint: %v", err)
+		}
+	} else {
+		n, err := note.Open(wcpRaw, note.VerifierList(w.Verifier))
+		if err != nil {
+			glog.Exitf("Failed to open CP: %v", err)
+		}
+
+		_, err = wcp.Unmarshal([]byte(n.Text))
+		if err != nil {
+			glog.Exitf("Failed to unmarshal CP: %v", err)
+		}
+	}
+
+	feeder := feeder{
+		wcp: wcp,
+		sdb: sdb,
+		w:   w,
+	}
+
+	tik := time.NewTicker(*pollInterval)
+	for {
+		glog.V(2).Infof("Tick: start feedOnce (witness size %d)", feeder.wcp.Size)
+		if err := feeder.feedOnce(ctx); err != nil {
+			glog.Exitf("Failed to feed: %v", err)
+		}
+		glog.V(2).Infof("Tick: feedOnce complete (witness size %d)", feeder.wcp.Size)
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-tik.C:
+		}
+	}
+}
+
+// feeder encapsulates the main logic and state of the feeder.
+// The residual logic outside of this should simply be initialization and
+// orchestration with a timer.
+// Note that feeder maintains its own state that represents the witness state.
+// This is optimized for the case where this is the only feeder for this log to
+// the witness. If the witness state is updated outwith this feeder, then the
+// number of successful updates from this feeder will drop. On the other hand,
+// if this is the only feeder then it avoids making requests to get the latest
+// checkpoint from the witness when it isn't changing.
+type feeder struct {
+	wcp *log.Checkpoint
+	sdb *audit.SumDBClient
+	w   http.Witness
+}
+
+// feedOnce gets the latest checkpoint from the SumDB server, and if this is more
+// recent than the witness state then it will construct a compact range proof by
+// requesting the minimal set of tiles, and then update the witness with the new
+// checkpoint. Finally, it will update the feeder's view of the witness state.
+func (f *feeder) feedOnce(ctx context.Context) error {
+	sdbcp, err := f.sdb.LatestCheckpoint()
+	if err != nil {
+		return fmt.Errorf("failed to get latest checkpoint: %v", err)
+	}
+	if int64(f.wcp.Size) >= sdbcp.N {
+		glog.V(1).Infof("Witness size %d >= SumDB size %d - nothing to do", f.wcp.Size, sdbcp.N)
+		return nil
+	}
+
+	glog.Infof("Updating witness from size %d to %d", f.wcp.Size, sdbcp.N)
+
+	broker := newTileBroker(sdbcp.N, f.sdb.TileHashes)
+
+	required := compact.RangeNodes(f.wcp.Size, uint64(sdbcp.N))
+	proof := make([][]byte, 0, len(required))
+	for _, n := range required {
+		i, r := convertToSumDBTiles(n)
+		t, err := broker.tile(i)
+		if err != nil {
+			return fmt.Errorf("failed to lookup %s: %v", i, err)
+		}
+		h := t.hash(r)
+		proof = append(proof, h)
+	}
+
+	wcpRaw, err := f.w.Update(ctx, *logID, sdbcp.Raw, proof)
+
+	if err != nil && !errors.Is(err, http.ErrCheckpointTooOld) {
+		return fmt.Errorf("failed to update checkpoint: %v", err)
+	}
+
+	// An optimization would be to immediately retry the Update if we get
```

Comment:
Could you put this as an explicit TODO?  In particular if this is run in a multi-feeder scenario then it seems very possible that the feeder will fail quite frequently (potentially always?) if it waits until the next new checkpoint to retry.

Reply:
I initially had this as an explicit TODO but decided not to for 2 reasons:
 * I currently have no intention of doing this
 * The primary intention for this feeder is to feed a local witness; this is documented on the `feeder` above (This is optimized for the case where this is the only feeder for this log to the witness.)

I leave this message here as a clear signpost for anyone that fancies changing that, but I'm actively working on reducing the number of TODOs to only those we would actually do given a bit of spare time. The trigger for doing this wouldn't be spare time, but the intention to run this as a collection of feeders to a single witness.

Leaving the comment open; resolve this if you're convinced, or holler back if you're not and I'll cancel my TODO-reduction thread and just make it so.